### ISSUE TITLE
FAT32 support

### ISSUE THREAD

**Issue by Wengier** [Wengier/dosbox-x-wiki#5]
DOSBox-X already supports the FAT32 file system directly. So some pages (such as the guides to install Windows 95/98) need to be updated accordingly as well to reflect this.

**Comment by rderooy** [comment]
I am planning on that, was just waiting to see if imgmake would be updated for fat32 support. Without that, there are additional steps needed to create the image... And the difference between what is now documented would not differ much.

WIth fat32 imgmake support, it gets much, much easier.

**Comment by Wengier** [comment]
@rderooy I am glad to see FAT32 IMGMAKE support too, but at this time I am not sure when @joncampbell123 plans to add it to DOSBox-X.

**Comment by rderooy** [comment]
Ok, if your fairly certain fat32 imgmake support will not make it before the next release, I will update the Win9x guides accordingly.

**Comment by Wengier** [comment]
It is already 24th of May. Since according to DOSBox-X’s official release pattern there will be no new features added 6 days before the end of the month, but only bug fixes, I think it probably won't be added to the 0.83.2 release.

**Comment by rderooy** [comment]
Ok, I intend to update the guides tomorrow, and perhaps we can get the guides added to the DOSBox-X wiki afterwards.

**Comment by Wengier** [comment]
Yes, we could do final cleanups of the Wiki pages in the next few days, and then ask @joncampbell123 to merge the content to the official site.

**Comment by joncampbell123** [comment]
Correct, IMGMAKE improvements will wait for the next release. The current code is actually pretty hacky and needs improvement overall.

**Comment by rderooy** [comment]
@Wengier I have updated the Win9x guides.

I think that once imgmake has been updated with FAT32 support, the instructions can be simplified, and the two could be consolidated into a single page, as there is a lot of duplication right now.

**Comment by Wengier** [comment]
@rderooy Thanks for the updates. I changed the following sentence from both pages from:

```According to Microsoft the official maximum size is 32GB, but in actuality FAT32 can support up to 2TB.```

To:

```While Windows 2000 or later won't let you format a volume bigger than 32GB with FAT32 using its built-in formatting tool, FAT32 can support up to 2TB as the maximum size.```

Because 32GB is **never** the official maximum size of FAT32 file system according to Microsoft. Windows 9x/Me officially supports partitions larger than 32GB when using the FAT32 file system. It is just that Windows 2000 or later won't let you format a volume bigger than 32GB with FAT32 using its built-in formatting tool. This is simply a limitation of Windows 2000 or later's built-in formatting tool ("by design" as stated in the webpage), and it does not make 32GB the official maximum size of FAT32 file system according to Microsoft.

**Comment by Wengier** [comment]
@rderooy Also, why did you revert all my recent edits in the page Guide:DOS Installation in DOSBox‐X when you added some new content to the page? I did not see any explainations for the revert anywhere.

**Comment by joncampbell123** [comment]
I have successfully used the code in datamtoolbox2 and Linux tools to format a 2TB FAT32 partition that Windows 10 is happy with. Of course that means a cluster size of 32KB which is not ideal.

Windows 98 setup however seems to have issues (it might even crash!) with FAT32 drives larger than about 64GB when it checks the disk, at least in DOSBox-X.

**Comment by Wengier** [comment]
@joncampbell123 Sure, some components of Windows 98 (FDISK too) may have issues with FAT32 drives larger than 64GB without any patch. But I believe there are official patches to fix most (if not all) of these issues. On the other hand, I don't think any of these issues are related to the 32GB limit of Windows 2000 or later's built-in formatting tool (which is by and large artificial).

**Comment by rderooy** [comment]
@Wengier Sorry about any revert, that was not intentional. I simply did not noticed that you had edited the page when I pasted a "new" version.

I don't work online, I use an offline editor and then apply the update.

Also I tried to clarify the Win2k thing, as it seemed out of place since the page is about win98.

**Comment by Wengier** [comment]
@rderooy Thanks for letting me know the reason. Please be careful then. For myself I always work online for any Wiki systems (this one or elsewhere), thus able to avoid such issue when there are more than one contributors. This can indeed happen when you work offline, in this case please try to check if there are updates by other contributors before applying the update. Thanks!

**Comment by Wengier** [comment]
@joncampbell123 I noticed the newly-created page [Guide:Managing image files in DOSBox‐X](../wiki/Guide%3AManaging-image-files-in-DOSBox‐X) by @rderooy, which is certainly pretty useful. At the same time, within the page I did see the following sentences:

```
Harddisk images greater than 2GB, up to 8GB will not be partitioned and formatted.
This is because FAT16 only supports partitions up to 2GB,
and larger partitions require FAT32 which is not supported by IMGMAKE at this point.
```

This is unfortunately still true at this time. If possible, could you please add FAT32 support for IMGMAKE before 0.83.3 is released? Thanks!

**Comment by joncampbell123** [comment]
IMGMAKE will have FAT32 support, yes. Haven't gotten around to it yet.

**Comment by Wengier** [comment]
@joncampbell123 I see. By the way, I want to mention that I think [pixel-perfect scaling](https://www.vogons.org/viewtopic.php?f=41&t=49160) is probably a very useful addition to DOSBox-X, so I sent a message to its author earlier today to invite him to add pixel-perfect scaling to DOSBox-X too. He agreed, although he said he may not do it very soon because his spare time is currently limited. In any case, he will likely submit a pull request for this later.

**Comment by rderooy** [comment]
Another patch that may be of interest, is the autotype function
https://github.com/dosbox-staging/dosbox-staging/commit/239396fec83dbba6a1eb1a0f4461f4a427d2be38

And I'm not sure the munt implementation in DOSBox-X is upto-date.

**Comment by Wengier** [comment]
@rderooy Yes, I am working on adding the AUTOTYPE command right now. Should be ready later.

**Comment by Wengier** [comment]
I have already added the AUTOTYPE command. I will also check the function to mount overlay drives.

**Comment by Wengier** [comment]
Now I added support for mounting overlay drives to DOSBox-X too. With the exception of pixel perfect scaling, at this time almost all features previously unique to DOSBox ECE or DOSBox-staging are supported by DOSBox-X too.

**Comment by Wengier** [comment]
@joncampbell123 and @rderooy The DOSBox-staging maintainers had invited me to contribute to the DOSBox-staging project too. I noticed that the two projects are indeed pretty similar at this time, and many features are welcomed by both projects. But contributing to both projects can lead to a dilemma of having to cut time in half, and the fact that one need to do one piece of work for both is always a great pain, not mentioning the issues of code split and the needs to port features from each other. So I suggested a possible merge of the projects, and they are considering it right now and may post an issue in the GitHub page later. Basically most things should be kept as is, but the code review system from DOSBox-staging may be introduced for cleaner source code.

**Comment by rderooy** [comment]
@wengier I had suggested this a while back. But I think merging them will
not be easy.

I would suggest, assuming everyone agrees, to start with a kind of
"inventory" first. What features, functions and bug fixes are missing from
one or the other with links to the commits.

**Comment by rderooy** [comment]
@Wengier @joncampbell123 in case you want to get hold of me outside github, you can email me at robert.de.rooy at google's gmail.com. I will delete this post after commenting as I assume both of you will get an email notification.

**Comment by rderooy** [comment]
A few of the things I think that need deciding on, assuming all agree, are
- Which fork will serve as the base
- Naming. Personally I would suggest to use the DOSBox-X name as the DOSBox-staging name has issues. And inventing a new name seems like a pointless exercise.
- Who has commit access. Do both @joncampbell123  and @dreamer have commit access, or does one take the lead. In any case, this is something to resolve between @joncampbell123  and @dreamer.

There are a few potential issues of conflict also. What about support for old Windows versions? The modernisation path that dosbox-staging has gone through is laudable, and I like how comparatively clean things are with the C++14 push and the cleaner build system. But it also means that dosbox-staging has issues on older systems.

Another issue is SDL1 vs SDL2. dosbox-staging has dropped support for SDL1, while dosbox-x has maintained it. Expanding it such that it works well enough in most cases.

**Comment by joncampbell123** [comment]
I suggest merging any feature that you like from DOSBox-X into staging. The modifications to DOSBox-X over the past 8 years have been extensive and probably differ a lot from your changes.

Pick the features you like and port from DOSBox-X one step at a time. Though I don't have a whole lot of free time right now, I can help port to DOSBox staging. Just let me know what you want so I can port it if I can.

DOSBox-X has targeted C++11 since 2017 so should port to C++14 easily enough.

@joncampbell123 (myself) has commit access.

**Comment by Wengier** [comment]
@joncampbell123 and @rderooy I have been discussing with them (primarily kcgen) since the last day, and at this time I think indeed there are a few potential issues for merging. So I will be neutral on the merging proposal as a whole myself, and more in favor towards the no direction.

**Comment by Wengier** [comment]
@joncampbell123 I don’t quite like DOSBox-staging’s code review system, even though I am fine with most other principles of their project, so I think I will still focus on the DOSBox-X project. Considering the potential issues mentioned with the merging process so far, I have largely dismissed the proposal myself.

**Comment by dreamer** [comment]
Just to be clear: requests in this thread did not come from dosbox-staging team (me and @kcgen ATM).

@Wengier @rderooy 
If you want to see some features from dosbox-staging in DOSBox-X - pick the ones you like and keep porting them (*but please, when porting our code to DOSBox-X, please keep our copyright intact and do credit original author of the code you copy - there are multiple mechanisms in Git to accomplish that - so far you ported code from us and sometimes removed all information about original authors - this is not cool*).

If you want to see some features from DOSBox-X in dosbox-staging - apply @joncampbell123 advice above. If you plan to stick for some time, **I am ok with giving out write access to dosbox-staging repo** (but it does not mean you can suddenly disregard code review/gating process - tell me if you're interested). If there's something unclear about our development process - please create a bug report in **dosbox-staging** so we could discuss it and improve the documentation (or, you can even suggest changes to the development process by sending PR to [CONTRIBUTING.md](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) document).

About merging the projects: we had this discussion once already; if there's a renewed interest in teaming up, then I think this time it should be discussed in DOSBox-X bugtracker, with clear confirmation from @joncampbell123 that there's interest. I believe long-term goals of our projects are not compatible, but perhaps I am wrong (I am often wrong, please correct me when I am :) ).

From my POV we are in quite comfortable position already - we share expertise and it's still possible to port some fixes between the projects - I don't consider our two projects to be competing, but rather indirectly supporting each other. Maybe it's even healthier for the community to have two projects focusing on their own goals.

*If there's interest in teaming up, then I can describe my idea on how this process could possibly go forward - ping me if you're interested because I would rather spend my time on writing code instead ;).*

**Comment by joncampbell123** [comment]
_From my POV we are in quite comfortable position already - we share expertise and it's still possible to port some fixes between the projects - I don't consider our two projects to be competing, but rather indirectly supporting each other. Maybe it's even healthier for the community to have two projects focusing on their own goals._

Exactly my concern. There are some additional changes I intend to make in the future, including the debugger and how the BIOS and DOS are stored and handled, that will make DOSBox-X's design drift further from it's original DOSBox SVN roots and architecture. The changes are intended to improve the debugger's functionality and to make BIOS and DOS interrupt handling more modular, realistic, and better able to cope with the guest application's virtual 8086 mode and paging.

However one thing that dosbox-staging might consider borrowing from DOSBox-X is the modified normal core (and possibly the 8086 and 286 normal cores as well). Modifications to the normal core have been made so that guest page faults are handled with a C++ exception rather than recursion into another emulator loop as SVN and other forks do (the recursion behavior is kept only for page faults that occur during a callback instruction). The reason this matters is that the modified behavior makes DOSBox-X work properly with preemptive multitasking OSes like Windows 95 where a task switch can happen during a page fault (and Windows 95 doesn't necessarily return in the same order---this is where DOSBox SVN and others start to fall apart). The other key modification to normal core is careful rearrangement of the code to make sure that each instruction either completes fully or leaves the CPU registers as it was if interrupted by a page fault (another cause of Windows 95 instability in other forks) so that it can resume properly when Window 95 pages in the memory. This part has been tested extensively by deliberately booting Windows 95 with as little as 4MB to make sure page faulting/swapfile use is as intense as it can be and running random applications.

Pulling in the normal core will also bring in the STI/CLI fix to reflect actual processor behavior. If interrupts are disabled, a STI immediately followed by a CLI does not cause interrupt processing (STI, NOP, CLI will however). It was a fix to improve emulation and address a particular PC-98 game that used STI+CLI in it's non-reentrant interrupt handler for some strange reason. The game never crashed on real hardware, but used to crash in DOSBox-X, because it's interrupt handler saves and restores variables in a way that cannot support reentrancy. For whatever reason, DOSBox SVN and other forks like to *immediately* process interrupts when processing STI and that's where the trouble happens.

**Comment by Wengier** [comment]
@dreamer I appreciate your comment and the point about the source code credit and copyright. It appears that the relevant sections in DOSBox-X's README about source code credit has not been updated for some time. I did make some update to it recently, but it is still very out of date. I will surely make further cleanups to it (and other relevant files) to properly credit the source code borrowed from DOSBox-staging (and other projects).

**Comment by Wengier** [comment]
@dreamer Perhaps I can first try to port features such as LFN support to DOSBox-staging, which we had discussed earlier. More features will be on the way, but one step at a time as advised by @joncampbell123.

**Comment by kcgen** [comment]
@Wengier , when I pull in other people's source, the first thing I do is commit it with authorship retained, using the `--author=` argument. 

I use git-aliases so I don't have to remember these details:

``` text
[alias]
    assb = commit --author='Sean Barrett <sean+github@xyz.xyz>'
    asdr = commit --author='David Reid <mackron@xyz.xyz>'
    asyc = commit --author='Yann Collet <yann.collet.73@xyz.xyz>'
    asjc = commit --author='Jonathan Campbell <jonathan@xyz.xyz>'
```

For example, to pull in DOSBox-X's paratrooper fix: `git asjc -m"Add fix for Paratrooper audio"`

This is how the commit appears, without any effort by @joncampbell123 needed:
![Screenshot at 2020-06-20 11-07-23](https://user-images.githubusercontent.com/1557255/85208544-4353a980-b2e6-11ea-9f48-22e27e4f5b86.png)

Then if any adjusts are needed, I do those in follow up commits, with just myself as the author.

Then in our Project's changelog, we would say:
- ` "Paratrooper audio fix - authored by joncampbell123 merged by kcgen"`, as opposed to
- `"Paratrooper audio fix (kcgen)"`, which would incorrectly ascribe myself as the sole author for this work.

We would like you to use `--author` and this textual change description, to ensure authorship history retained and proper credit is given.

Likewise, if you take works that come in their own files (like AUTOTYPE did), the header and copyright lines must be retained and not be stripped away like you've done to AUTOTYPE.   If you must lift out the code and place it in a larger DOSBox-X file, then you can transplant the copyright history into the augmented file, and point out which sections the copyright applies to.

**Comment by Wengier** [comment]
@kcgen I usually use the GitHub web interface to commit files for DOSBox-X so far, and I have not used the command-line tool often for the project, so I never knew the existence of the --author= argument. Perhaps this feature is only supported by the command-line tool but not the web interface? In any case, thank your for letting me aware of this.

**Comment by dreamer** [comment]
@Wengier Yes, it is only supported by command-line tool. It is very hard to maintain our way of work without using CLI git client (we depend heavily of amending commits for PR and interactive rebase to clean them up). I can share or write some tutorials explainig how to do specific tasks using CLI client if there's such need.

The way @kcgen described is one way of dealing with preserving code authorship info, but there are others as well:

- You can use standard Git's `Signed-off` trailer to the commit authored by someone else (`git commit --signoff` appends this trailer automatically).
- GitHub supports co-authorship info via [`Co-authored-by` trailer](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) - but it's a GitHub feature, not Git feature.

(I don't know which one is the preferred way in DOSBox-X)

**Comment by kcgen** [comment]
Co-authored produces this in the Git log:

``` text
Commit message
Co-authored-by: name <name@example.com>
Co-authored-by: another-name <another-name@example.com>
```

This makes sense if you deeply reworked the original code (say > 50% ) or if you truly collaborated on the solution or implementation together.

Another litmus test is to ask yourself: _"would the other author agree that what I contributed justifies also being an author?"_  If yes.. then go for it!

I certainly wouldn't say that I co-authored the Paratrooper fix, for example.

**Comment by Wengier** [comment]
@kcgen I think I have an example here: for the LS command, I did initially ported the command from DOSBox-staging, but I deeply reworked the original code to make it support long file names and some options. The change was so great (>50%) that I think I perhaps should be justified also being an author in this case.

**Comment by Wengier** [comment]
@dreamer Thanks. I will try to use the Git command-line tool more later on for the project.

**Comment by dreamer** [comment]
@joncampbell123 

> (…) There are some additional changes I intend to make in the future, including the debugger and how the BIOS and DOS are stored and handled, that will make DOSBox-X's design drift further from it's original DOSBox SVN roots and architecture. The changes are intended to improve the debugger's functionality and to make BIOS and DOS interrupt handling more modular, realistic, and better able to cope with the guest application's virtual 8086 mode and paging.

Yup, we're drifting away from SVN as well, but we're very careful about what changes from SVN we don't want (and revert them before merging new version of svn/trunk branch). But compared to DOSBox-X, we're very close to upstream still.

The big changes, that we plan (that might not be compatible with DOSBox-X) are: 

- completely replacing the build system to make it feasible to introduce unit tests and integration tests (we could introduce unit testing right now, but it would cement the autoconf in place, and we really want to get rid of it).
- turning dosbox-staging from mixed C/C++ program to mixed C/C++/Rust program (we would prefer to go for Rust rather than C++20)
- more automation, more static analysis, very specific solutions to prevent certain classes of regressions
- very opinionated GUI changes
- perhaps, if inclusion of Rust will give us benefits we hope are possible: removal/rewrite of large chunks of old DOSBox code (e.g. replacement of SDL_net dependency with Rust standard library, removal of low-quality OPL emulators).
- some other changes, that I prefer to keep under wraps for now ;)

And we have some long-term plans, where we hope on very close cooperation, e.g.:

- Introduction of ROM-like format for distributing DOS software. *Imagine the possibility of distributing a game or demo using single file - end-user would only double-click on the ROM-like file and dosbox implementation would take care of safely mounting the content and preparing directory overlays to store save files.*

    When we'll be ready to start work on ROM-like format, my intention is to start by gathering requirements from interested parties and by creating [RFC](https://en.wikipedia.org/wiki/Request_for_Comments) document/technical format specification. Hopefully, we can cooperate on the specification (maybe on reference implementation as well), so the new file format would be supported by both projects from the get-go.

> However one thing that dosbox-staging might consider borrowing from DOSBox-X is the modified normal core (and possibly the 8086 and 286 normal cores as well). (…) The reason this matters is that the modified behavior makes DOSBox-X work properly with preemptive multitasking OSes like Windows 95 where a task switch can happen during a page fault (and Windows 95 doesn't necessarily return in the same order---this is where DOSBox SVN and others start to fall apart).

Just to be clear - are you talking about running emulator using Windows 95 or Windows 95 in an emulator? Because the statement can be read both ways :)

I am not aware of any problems resulting from usage of normal cores on Linux (which is preemptive multitasking kernel, and AFAIK tasks switches can happen during page fault on Linux). We'll look into it, but there's not many games that require usage of `normal` core any more - most software is well serviced by either one of 2 dynamic recompiler cores we inherited from SVN.

Regardless, we'll look into improvements to normal core. PC-98 support is not on our agenda right now, but we would like to support it at some point though - we'll definitely ping you when starting work on it :)

To reciprocate, DOSBox-X might consider:

- take a look at later changes to `src/libs/decoders` - you are missing security fixes, opus support, perhaps SDL2 compatibility fixes (not sure)
- we recently introduced our own version of `byteswap.h` header (it's a bit more focused than header with the same name used by DOSBox-X) and set of functions to address rampant memory alignment issues plaguing our upstream see `mem_host.h`, `mem_unaligned.h`
- pixel-perfect mode is quite easy to re-implement from scratch (I am not sure about quirks resulting from SDL1 usage though).
- you probably want to invest time into porting new dynamic recompiler code from SVN

**Comment by dreamer** [comment]
> @dreamer Thanks. I will try to use the Git command-line tool more later on for the project.

@Wengier If you'll have any questions (how to accomplish a specific task for example), then feel free to drop me a mail or create a "question" issue in the bugtracker - Git has plenty of features and users often don't know about some really cool ones :)

**Comment by Wengier** [comment]
@kcgen I have thought about your point about the copyright header for AUTOTYPE command. Indeed when I added new features I usually only inserted code and some inline-comments, but not necessarily inserting any headers in the source code itself. It does not only apply to AUTOTYPE command borrowed from DOSBox-staging, but *any* commands added by myself too. For example, when I implemented new commands (ATTRIB, BREAK, COUNTRY, FOR, and many more) myself, I never added a header to point out that I wrote this command in the code. I would prefer to add the credit to some external document later.

**Comment by Wengier** [comment]
@dreamer Sure. I will first try to get myself more familiar with the Git command in general, and if have any more questions, I will ask you about it. Thanks!

**Comment by kcgen** [comment]
@Wengier , regarding copyright headers: _"I would prefer to add the credit to some external document later."_

The GPL requires the copyright header be retained in full. So your suggestion of externalizing the copyright line won't work. (That would allow either deliberate or accidental plagarism by an external party).

There are good examples of source code with a long pedigree of open source contributors over decades (if you read sources for the GNU tools or say emacs) . 

In our case of staging and X, all our source is GPL2, so we've got that license in place. 

What we then do is keep a running tab of the copyright owners at the top of the file. 

So for lifting out the AUTOTYPE code into your larger existing source file, you would take our copyright line and add it at the top of your existing list of copyright lines as follows:

`AUTOTYPE Copyright (C) 2020-2020  The dosbox-staging team`

For your own additions:

`ATTRIB, BREAK, COUNTRY, and FOR Copyright (C) 2020-2020  The DOSBox-X team.`

This way, a single source file can become an amalgamation of GPL-compatible contributions without losing the copyright history for the parts.

**Comment by joncampbell123** [comment]
_Introduction of ROM-like format for distributing DOS software._

@dreamer I suggest using the ZIP file format as the storage container for other elements of the ROM-like format. ZIP is pretty easy to implement, and it would allow optional compression of the contents. Then inside the ZIP, lay out the specific filesystem structure the emulators will use to determine how to run the game.

Some forks if I recall correctly can already mount ZIP files as a drive letter. The "ROM" format would just have certain files in specific places that define how to run it.

What would be a good initial feature set for the ROM format to have? Would the ROM format have it's own AUTOEXEC/CONFIG sections and dosbox.conf inside?

The ZIP format could also contain subdirectories that the dosbox.conf could mount as drive letters, or it could also contain disk images that are IMGMOUNTed or BOOTed as needed.

Use of the ZIP format would also make image creation easy since almost anything out there can create and extract ZIP files. A Windows user could use 7-zip, for example. Here on Linux, the InfoZip utility could be used.

**Comment by Wengier** [comment]
@kcgen I have already added a new CREDITS.md page for updated credits info and added copyright headers for DOS commands as required by GNU GPL in my latest pull request.

**Comment by kcgen** [comment]
Thank you @Wengier!

**Comment by Wengier** [comment]
@dreamer I have now updated the decoder files in src/libs/decoders to the latest version in my latest pull request. Thanks for pointing that out. Now I will look at the Git command-line tool and feature porting to DOSBox-staging.

**Comment by Wengier** [comment]
@dreamer Of course I'm happy to contribute to DOSBox-staging because of your invitation. Even though it is not the preferred code review system for me, I can try to get used to it.

**Comment by rderooy** [comment]
Regarding the ROM-like format. I think that would be a great idea. Younger folks don't know DOS commands, and similar to ROM files expect things to just work.

I think using ZIP would be fine, but I think it should have a different extension such that;
1. Similar to Java JAR files, to make it easier to recognise what it is meant for
2. When the user double clicks on them, they don't open in 7zip, winzip, winrar, whatever, but in DOSBox.
3. Prevent people unpacking it, and then complaining about not knowing what to do with it.

No need to stay with a 3 letter extension I think, especially since most 3 letter combinations have already been used. e.g. DBE, DBX, DBZ are all already used by other programs.

And as @joncampbell123 already mentioned, I think using a folder structure in the ZIP like this should work for folder mounts:

\dosbox.conf
\a.fd\diskette files
\c.hd\hdd files
\d.cd\cdrom files

For disk images it could be something like this:
\dosbox.conf
\a.fd.ima
\c.hd.img
\d.cd.iso

Or you could just assume that the dosbox.conf file has the right (relative) mount commands...

Where it gets tricky is sound and video options. Do you decide for the end-user which sound and video options the game will be setup with, or do you want to present a choice? This also ties into what the emulator is able to do. For instance while DOSBox-X and DOSBox ECE has support for MT-32 and General MIDI, this support is not present, or only for some platforms in other DOSBox forks.

Part of this may be resolved by having multiple config files, targeting different DOSBox forks, or different video and audio options. But it gets messy very quickly. Also because while a DOSBox fork may not support some option now, it may support it tomorrow, and you don't want to have to go and update the "rom" files. e.g.
\default.conf
\dosbox-x.conf
\vga.mt32.conf
\vga.sb.conf
\tga.conf

To support save files, I suggest using overlayfs support as present in DOSBox SVN, and most other DOSBox forks by now. Although that will not work for image files, unless you implement something like qcow2.

**Comment by dreamer** [comment]
We haven't started work on ROM-like format, but me and @kcgen believe ZIP is not going to be the best solution.

@rderooy I completely agree - we need new file extension and a mimetype; that's why I want to start with RFC document, to agree on some baseline solution so several projects can aim for the same goal (similarly like music or video container formats work) :)

Instead of ZIP we would prefer to use .ZSTD (as it is much better suited - bettern compression ratio and better decompression speed); also, format can't be a single compressed file (there are several technical issues that would cause - one being double-compression on audio files). I'll expand on the solution I suggest once I'll start writing RFC.

> Where it gets tricky is sound and video options. Do you decide for the end-user which sound and video options the game will be setup with, or do you want to present a choice?

It depends on what option exactly do we talk about. `output = something`? (definitely should not be allowed inside ROM-like file - it's user-selectable option); support for switching between different game-specific solutions (MIDI vs MT-32 vs SB for example) - that's ultimately for a packager to decide, but we *need* to have a solution for a packager to provide this choice to the end-user.

In Boxtron I solved similar problem this way: "MIDI autoconfiguration is supported by default in Boxtron; user can disable it. If MIDI is enabled AND if game supports MIDI (Boxtron provides tiny database of few games that do), then reconfigure game to use MIDI; else reconfigure game to use default sound settings." Works really nice in practice (users don't need to learn/switch anything themselves - it "just works").

> To support save files, I suggest using overlayfs support (…)

I think you mean "overlay mounts" (`overlayfs` is a Linux thing, not DOSBox thing) :) Yes; we can't have ROM-like files and write back to the files inside. We already made some changes to allow for mounting read-only files (something that was missing from SVN).

Overall there are many non-trivial edge-cases - (e.g. support for games using multiple CDs will require implementations to be able to read only certain file blocks, not the whole package upfront) - just enough to warrant a new container format; in the past I've seen 2-3 "solutions" based on simply zipping the files - none of them stuck because people pushing for those solutions did not consider use-cases needed by different people (or even users of a different OS).

**Comment by Wengier** [comment]
@rderooy You mentioned "overlayfs", if by this term you meant "overlay mounts" as pointed out by @dreamer, then DOSBox-X supports it too. In fact, in DOSBox-X overlay mounts supports additional features including long filename support which are not present in DOSBox SVN and other forks.

**Comment by rderooy** [comment]
Yes, I did mean overlay mounts in this regard.

**Comment by Wengier** [comment]
@joncampbell123 Some significant progresses had been made in recent days, which is certainly a great thing. However, I am also aware that there may be a misunderstanding that I had at some point suggested a possible merge of the two projects simply because I wanted to see some features in DOSBox-staging and I thought about my own time may be cut in half. Of course this was **not** how it started. Basically this is what had happened:

About 4 months ago @dreamer contacted me and sincerely invited me to port some features such as long filename support and mouse copy and paste from DOSBox LFN (my own DOSBox fork at that time) to DOSBox-staging and asked me to contribute to the DOSBox-staging project. I was initially very glad to see the appearance of the DOSBox-staging project, but knowing some more details of its code review system as he told me, after some thoughts I decided to **respectfully decline**, and the reason was exactly because I did not quite like DOSBox-staging's code review system. After that I started to contribute to the DOSbox-X project and worked on it regularly. Then I was in touch with @kcgen in DOSBox-X's issue tracker and I started to think about @dreamer's invitation again. Considering the recent developments of the two projects and the fact that the they had become even more similar than before (with more common features etc), as well as the fact that the features @dreamer invited me to port to DOSBox-staging were already ported to DOSBox-X in recent months, I started to suggest a possible merge of the projects in order to avoid the fact of putting one work for both projects and any future need of porting features to each other.

To summarize, it is not that I wanted to see some features in DOSBox-staging and I simply thought about my own time so I had at one point suggested a merge (which I fully dismissed after some discussion and considering the potential issues). Instead, I am very happy to port some features to DOSBox-staging because of @dreamer's invitation (as he specifically asked me to port some features earlier), but the main obstacle for me was always its code review system, as happened earlier and in recent days. Considering DOSBox-staging's code review system does have its own benefits (as @kcgen explained to me in details recently by email), so I decided to re-consider the idea of porting features to DOSBox-staging, overcoming any issue with its code review system by trying to get used to it. I will still focus on the DOSBox-X project, but I am glad to port any features invited by @dreamer or @kcgen (either previously or in the future) to DOSBox-staging.

**Comment by rderooy** [comment]
@joncampbell123 I have updated the Guides for the IMGMAKE FAT32 support, in addition to various other updates. If @Wengier has no issues, could you pull the changes?

**Comment by Wengier** [comment]
> @joncampbell123 I have updated the Guides for the IMGMAKE FAT32 support, in addition to various other updates. If @Wengier has no issues, could you pull the changes?

@joncampbell123 and @rderooy I am also fine with the latest Wiki contents. They are already targeted at the recently-released DOSBox-X 0.83.3, so there should be no issue to pull the changes.
